### PR TITLE
#5097: [TEST] Add "protected" to available paths api

### PR DESCRIPTION
**Pull request by pkazlenka**
* Added new tests in PathSpec
* Refactored existing PathSpec tests
* New class SwitchPairFilter to store named filters for switch pairs

### CODE REVIEW

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/network/PathsSpec.groovy [hunk 1]
```diff
@@ -142,28 +137,86 @@ class PathsSpec extends HealthCheckSpecification {
...
+        def paths = northbound.getPaths("convert switch pair to paths request parameters"(switchPair) +
+                [(PROTECTED): isIncluded])
+
+        then: "All paths #isIncludedString have protected ones and protected path differs from main path"
+        paths.findAll(verificationFilter).isEmpty()
+
+        where:
+        isIncludedString | isIncluded | verificationFilter
+        ""               | true       | { it.getNodes() == it.protectedPath.getNodes() }
+        "not"            | false      | { it.getProtectedPath() != null }
```
[the remark this answers is not in the document]
Comment (pkazlenka) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/network/PathsSpec.groovy:185]: You are right. I've confused protected path with flows in diversity group (which can have overlapping segments). Will add a check.

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/network/PathsSpec.groovy [hunk 2]
```diff
@@ -142,28 +137,86 @@ class PathsSpec extends HealthCheckSpecification {
...
         }
     }
+
+    @Tidy
+    @Tags(LOW_PRIORITY)
+    @Unroll
+    def "Protected path is #isIncludedString included into path list if #isIncludedString requested"() {
+        given: "Two switches with potential protected path"
+        def switchPair = topologyHelper.allNotNeighboringSwitchPairs
+                .findAll { WITH_AT_LEAST_N_DIVERSE_PATHS(it, 2) }
```
[the remark this answers is not in the document]
Comment (pkazlenka) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/network/PathsSpec.groovy:171]: In this particular case 'diverse' means 'different' and doesn't have anything related to 'diverse group flows'. This is how the set up was described in our tests historically. But I agree, it's better to switch to something like 'WITH_AT_LEAST_N_NON_OVERLAPPING_PATHS'.

File: src-java/testing/test-library/src/main/java/org/openkilda/testing/service/northbound/NorthboundServiceImpl.java [hunk 3]
```diff
@@ -647,34 +648,15 @@ public List<FlowPayload> getSwitchFlows(SwitchId switchId, Integer port) {
     }
 
     @Override
-    public PathsDto getPaths(SwitchId srcSwitch, SwitchId dstSwitch, FlowEncapsulationType flowEncapsulationType,
-                             PathComputationStrategy pathComputationStrategy,
-                             Long maxLatency, Long maxLatencyTier2, Integer maxPathCount) {
+    public List<PathDto>  getPaths(Map<PathRequestParameter, Object> parameters) {
+        /* it looks like it makes mor sense to pass SwitchPair as the mandatory first parameter, but this leads to
```
[the remark this answers is not in the document]
Comment (pkazlenka) [reply src-java/testing/test-library/src/main/java/org/openkilda/testing/service/northbound/NorthboundServiceImpl.java]: Fixed

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/network/PathsSpec.groovy (hunk 2 above)
[the remark this answers is not in the document]
Comment (pkazlenka) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/network/PathsSpec.groovy]: Fixed

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/network/PathsSpec.groovy (hunk 1 above)
[the remark this answers is not in the document]
Comment (pkazlenka) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/network/PathsSpec.groovy:186]: Added better assertion.

**Review by dmitrii-beliakov — APPROVED** [review]

**Pull request closed without merging**